fvm/state/ledger_address_generator.go

```diff
@@ -0,0 +1,49 @@
+package state
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+)
+
+const keyAddressState = "account_address_state"
+
+// ledgerBoundAddressGenerator is a decorator for an address generator.
+// It uses the underlying generator it gets from the chain.
+// The only change is that when next address is called the state is updated as well.
+type ledgerBoundAddressGenerator struct {
+	generator flow.AddressGenerator
+	ledger    Ledger
+}
+
+func NewLedgerBoundAddressGenerator(ledger Ledger, chain flow.Chain) (flow.AddressGenerator, error) {
+	stateBytes, err := ledger.Get("", "", keyAddressState)
+	if err != nil {
+		return nil, err
+	}
+
+	addressGenerator := chain.BytesToAddressGenerator(stateBytes)
+	return &ledgerBoundAddressGenerator{
+		ledger:    ledger,
+		generator: addressGenerator,
+	}, nil
+}
+
+func (g *ledgerBoundAddressGenerator) NextAddress() (flow.Address, error) {
+	address, err := g.generator.NextAddress()
```

Comment:
I'm a little bit concerned about the fact that this relies on the internal state of `g.generator` staying in sync with the address state stored in `g.ledger`. This should be fine in theory (since no other component should be writing to this key) but I still worry that the dependency is a bit too implicit. 

You could avoid this by constructing a new generator instance for each call to `NextAddress`:

```go
stateBytes, err := ledger.Get("", "", keyAddressState)
if err != nil {
    return nil, err
}

address, err := g.chain.BytesToAddressGenerator(stateBytes).NextAddress()
if err != nil {
    return nil, err
}
```

In this case you'd probably want to replace `g.generator` with `g.chain`

Reply:
Creating a new generator every time would mean the interface for `flow.AddressGenerator.CurrentAddress()` and `Bytes()` would have to change to also return an error since getting the ledger state returns an error. 

After a discussion with Peter, we decided it can remain as it is now.